File: apps/scraper/src/scrapers/fivethirtyeight.py
"""
FiveThirtyEight Scraper

Scrapes polling data from FiveThirtyEight's polling database
"""

import asyncio
import json
from typing import List, Dict, Any
from datetime import datetime
from loguru import logger

from .base import BaseScraper


class FiveThirtyEightScraper(BaseScraper):
    """Scraper for FiveThirtyEight polling data"""

    BASE_URL = "https://projects.fivethirtyeight.com"
    POLLS_API = f"{BASE_URL}/polls-page/data/president_polls.json"

    def __init__(self):
        super().__init__("FiveThirtyEight")

    async def scrape(self) -> List[Dict[str, Any]]:
        """
        Scrape polls from FiveThirtyEight

        FiveThirtyEight provides JSON data which makes scraping cleaner

        Returns:
            List of poll dictionaries
        """
        logger.info("Scraping FiveThirtyEight polls")

        try:
            # Fetch the JSON data
            html = self.fetch_page(self.POLLS_API)
            data = json.loads(html)

            polls = []

            # 538 provides structured JSON data
            for poll_entry in data:
                try:
                    poll_data = self._parse_poll_entry(poll_entry)
                    if poll_data:
                        polls.append(poll_data)
                except Exception as e:
                    logger.warning(f"Failed to parse poll entry: {e}")
                    continue

            logger.success(f"Scraped {len(polls)} polls from FiveThirtyEight")
            return polls

        except Exception as e:
            logger.error(f"FiveThirtyEight scraping failed: {e}")
            raise

    def _parse_poll_entry(self, entry: Dict) -> Dict[str, Any]:
        """
        Parse a single poll entry from 538 JSON

        Args:
            entry: Poll entry dictionary

        Returns:
            Poll dictionary or None
        """
        # FiveThirtyEight JSON structure (example - actual structure may vary)
        # The real API structure should be inspected first

        try:
            # Extract basic info
            pollster = entry.get('pollster', 'Unknown')
            poll_id = entry.get('poll_id', '')
            start_date = entry.get('start_date', '')
            end_date = entry.get('end_date', '')
            sample_size = entry.get('sample_size')
            population = entry.get('population', 'lv')  # lv, rv, a
            methodology = entry.get('methodology', 'online')

            # Get race information
            race_name = entry.get('race', entry.get('state', 'National'))
            if race_name == 'National':
                race_name = 'Presidential'

            # Extract candidate results
            results = {}

            # 538 typically has 'answers' array with candidate data
            if 'answers' in entry:
                for answer in entry['answers']:
                    candidate = answer.get('choice', '')
                    percentage = answer.get('pct', 0)
                    if candidate and percentage:
                        results[candidate] = float(percentage)

            # Fallback: look for direct candidate fields
            elif 'candidates' in entry:
                for candidate in entry['candidates']:
                    name = candidate.get('name', '')
                    pct = candidate.get('pct', 0)
                    if name and pct:
                        results[name] = float(pct)

            if not results:
                logger.debug(f"No results found for poll {poll_id}")
                return None

            poll_data = {
                'source': 'FiveThirtyEight',
                'source_poll_id': str(poll_id),
                'race_name': race_name,
                'pollster_name': pollster,
                'poll_date': end_date or start_date,  # Use end date as poll date
                'start_date': start_date,
                'end_date': end_date,
                'sample_size': sample_size,
                'results': results,
                'methodology': self._map_methodology(methodology),
                'population_type': self._map_population(population),
                'pollster_rating': entry.get('fte_grade', ''),  # 538's pollster grade
                'scraped_at': datetime.now().isoformat(),
            }

            return poll_data

        except Exception as e:
            logger.debug(f"Failed to parse poll entry: {e}")
            return None

    def _map_methodology(self, methodology: str) -> str:
        """
        Map 538 methodology to our standard format

        Args:
            methodology: 538 methodology string

        Returns:
            Standard methodology string
        """
        method_lower = methodology.lower()

        if 'phone' in method_lower or 'live' in method_lower:
            return 'PHONE'
        elif 'online' in method_lower or 'web' in method_lower:
            return 'ONLINE'
        elif 'ivr' in method_lower:
            return 'IVR'
        elif 'mixed' in method_lower:
            return 'MIXED'
        else:
            return 'OTHER'

    def _map_population(self, population: str) -> str:
        """
        Map 538 population type to our standard format

        Args:
            population: 538 population string

        Returns:
            Standard population type
        """
        pop_lower = population.lower()

        if pop_lower == 'lv' or 'likely' in pop_lower:
            return 'LV'  # Likely Voters
        elif pop_lower == 'rv' or 'registered' in pop_lower:
            return 'RV'  # Registered Voters
        elif pop_lower == 'a' or 'adult' in pop_lower:
            return 'A'   # Adults
        elif pop_lower == 'v' or 'voter' in pop_lower:
            return 'V'   # Voters
        else:
            return 'A'   # Default to adults

    async def scrape_state_polls(self, state: str) -> List[Dict[str, Any]]:
        """
        Scrape polls for a specific state

        Args:
            state: State abbreviation (e.g., 'PA', 'GA')

        Returns:
            List of poll dictionaries for that state
        """
        logger.info(f"Scraping FiveThirtyEight polls for {state}")

        # 538 may have state-specific endpoints
        state_url = f"{self.BASE_URL}/polls-page/data/{state.lower()}_polls.json"

        try:
            html = self.fetch_page(state_url)
            data = json.loads(html)

            polls = []
            for entry in data:
                poll_data = self._parse_poll_entry(entry)
                if poll_data:
                    polls.append(poll_data)

            logger.success(f"Scraped {len(polls)} polls for {state}")
            return polls

        except Exception as e:
            logger.warning(f"Failed to scrape {state} polls: {e}")
            return []


async def main():
    """Run 538 scraper standalone"""
    logger.info("Starting FiveThirtyEight scraper")

    scraper = FiveThirtyEightScraper()
    try:
        result = await scraper.run()
        logger.info(f"Scraper result: {result}")
    finally:
        scraper.close()


if __name__ == "__main__":
    asyncio.run(main())

File: apps/api/src/scrapers/fivethirtyeight.ts
/**
 * FiveThirtyEight scraper
 * Scrapes polling data from projects.fivethirtyeight.com
 */

import * as cheerio from 'cheerio'
import { BaseScraper } from './base-scraper'
import { ScraperResult, ScrapedRace, ScrapedPoll } from './types'

export class FiveThirtyEightScraper extends BaseScraper {
  private readonly baseUrl = 'https://projects.fivethirtyeight.com'

  constructor() {
    super('FiveThirtyEight', {
      rateLimit: 3000, // Be extra respectful - 3 seconds between requests
    })
  }

  async scrape(): Promise<ScraperResult> {
    const result: ScraperResult = {
      success: false,
      races: [],
      errors: [],
      scrapedAt: new Date(),
      source: 'FiveThirtyEight',
    }

    try {
      this.log('Starting FiveThirtyEight scrape...')

      // FiveThirtyEight provides CSV/JSON data which is easier to parse
      // We'll fetch their polls data via their public API/CSV endpoints

      // Scrape presidential race
      const presidentialRace = await this.scrapePresidentialRace()
      if (presidentialRace) {
        result.races.push(presidentialRace)
      }

      // Scrape Senate races
      const senateRaces = await this.scrapeSenateRaces()
      result.races.push(...senateRaces)

      result.success = result.races.length > 0
      this.log(`Scrape completed: ${result.races.length} races, ${result.errors.length} errors`)

      return result
    } catch (error) {
      this.logError('Fatal error during scrape', error)
      result.errors.push((error as Error).message)
      return result
    }
  }

  private async scrapePresidentialRace(): Promise<ScrapedRace | null> {
    try {
      this.log('Scraping presidential race from FiveThirtyEight...')

      // FiveThirtyEight provides structured JSON/CSV data
      // We'll use their polls API endpoint
      const url = `${this.baseUrl}/polls-page/data/president_polls.csv`

      const response = await this.client.get(url)
      const csvData = response.data

      // Parse CSV data
      const polls = this.parseCSVPolls(csvData, 'president')

      this.log(`Found ${polls.length} presidential polls from 538`)

      if (polls.length === 0) {
        return null
      }

      return {
        slug: 'president-2024',
        name: '2024 Presidential Election',
        type: 'president',
        candidates: this.extractCandidates(polls),
        polls,
      }
    } catch (error) {
      this.logError('Error scraping 538 presidential race', error)
      return null
    }
  }

  private async scrapeSenateRaces(): Promise<ScrapedRace[]> {
    const races: ScrapedRace[] = []

    try {
      this.log('Scraping Senate races from FiveThirtyEight...')

      await this.sleep(this.config.rateLimit)

      const url = `${this.baseUrl}/polls-page/data/senate_polls.csv`
      const response = await this.client.get(url)
      const csvData = response.data

      // Parse and group by state
      const pollsByState = this.groupPollsByState(csvData)

      for (const [state, polls] of Object.entries(pollsByState)) {
        if (polls.length > 0) {
          const race: ScrapedRace = {
            slug: `senate-${state.toLowerCase()}-2024`,
            name: `${state} Senate 2024`,
            type: 'senate',
            state,
            candidates: this.extractCandidates(polls),
            polls,
          }
          races.push(race)
        }
      }

      this.log(`Found ${races.length} Senate races from 538`)
    } catch (error) {
      this.logError('Error scraping 538 Senate races', error)
    }

    return races
  }

  /**
   * Parse CSV polls data
   * FiveThirtyEight CSV format has columns like:
   * poll_id, pollster, sponsor, display_name, pollster_rating_id, start_date, end_date,
   * sample_size, population, methodology, office_type, seat_number, seat_name,
   * answer, candidate_name, pct
   */
  private parseCSVPolls(csvData: string, raceType: string): ScrapedPoll[] {
    const polls: ScrapedPoll[] = []
    const lines = csvData.split('\n')

    if (lines.length < 2) {
      return polls
    }

    // Parse header
    const headers = lines[0].split(',').map((h) => h.trim().replace(/"/g, ''))

    // Group rows by poll_id
    const pollsMap = new Map<string, any>()

    for (let i = 1; i < lines.length; i++) {
      const line = lines[i].trim()
      if (!line) continue

      try {
        const values = this.parseCSVLine(line)
        if (values.length < headers.length) continue

        const row: any = {}
        headers.forEach((header, index) => {
          row[header] = values[index]?.trim() || ''
        })

        const pollId = row.poll_id || row.question_id || `poll_${i}`

        if (!pollsMap.has(pollId)) {
          pollsMap.set(pollId, {
            pollId,
            pollster: row.pollster || row.display_name || 'Unknown',
            startDate: row.start_date || row.field_start,
            endDate: row.end_date || row.field_end,
            sampleSize: row.sample_size,
            methodology: row.methodology,
            population: row.population,
            results: {},
          })
        }

        const poll = pollsMap.get(pollId)
        const candidateName = row.candidate_name || row.answer
        const percentage = parseFloat(row.pct || row.percentage || '0')

        if (candidateName && !isNaN(percentage)) {
          poll.results[candidateName] = percentage
        }
      } catch (error) {
        // Skip malformed lines
        continue
      }
    }

    // Convert map to array of ScrapedPoll objects
    for (const pollData of pollsMap.values()) {
      if (Object.keys(pollData.results).length >= 2) {
        // Only include polls with at least 2 candidates
        const poll: ScrapedPoll = {
          pollster: pollData.pollster,
          pollsterSlug: this.normalizePollsterSlug(pollData.pollster),
          pollDate: this.parseDate(pollData.endDate) || new Date(),
          fieldDateStart: this.parseDate(pollData.startDate),
          fieldDateEnd: this.parseDate(pollData.endDate),
          sampleSize: this.parseNumber(pollData.sampleSize),
          methodology: this.mapMethodology(pollData.methodology),
          populationType: this.mapPopulation(pollData.population),
          results: pollData.results,
          sourceUrl: `${this.baseUrl}/polls/`,
        }

        polls.push(poll)
      }
    }

    return polls
  }

  /**
   * Parse CSV line handling quoted fields
   */
  private parseCSVLine(line: string): string[] {
    const result: string[] = []
    let current = ''
    let inQuotes = false

    for (let i = 0; i < line.length; i++) {
      const char = line[i]

      if (char === '"') {
        inQuotes = !inQuotes
      } else if (char === ',' && !inQuotes) {
        result.push(current)
        current = ''
      } else {
        current += char
      }
    }

    result.push(current)
    return result
  }

  /**
   * Group polls by state
   */
  private groupPollsByState(csvData: string): Record<string, ScrapedPoll[]> {
    const lines = csvData.split('\n')
    if (lines.length < 2) return {}

    const headers = lines[0].split(',').map((h) => h.trim().replace(/"/g, ''))
    const statePolls: Record<string, ScrapedPoll[]> = {}

    const pollsMap = new Map<string, any>()

    for (let i = 1; i < lines.length; i++) {
      const line = lines[i].trim()
      if (!line) continue

      try {
        const values = this.parseCSVLine(line)
        if (values.length < headers.length) continue

        const row: any = {}
        headers.forEach((header, index) => {
          row[header] = values[index]?.trim() || ''
        })

        const state = row.state || row.seat_name
        if (!state || state.length !== 2) continue

        const pollId = `${state}_${row.poll_id || row.question_id || i}`

        if (!pollsMap.has(pollId)) {
          pollsMap.set(pollId, {
            pollId,
            state,
            pollster: row.pollster || row.display_name || 'Unknown',
            startDate: row.start_date || row.field_start,
            endDate: row.end_date || row.field_end,
            sampleSize: row.sample_size,
            methodology: row.methodology,
            population: row.population,
            results: {},
          })
        }

        const poll = pollsMap.get(pollId)
        const candidateName = row.candidate_name || row.answer
        const percentage = parseFloat(row.pct || row.percentage || '0')

        if (candidateName && !isNaN(percentage)) {
          poll.results[candidateName] = percentage
        }
      } catch (error) {
        continue
      }
    }

    // Group by state
    for (const pollData of pollsMap.values()) {
      if (Object.keys(pollData.results).length >= 2) {
        const state = pollData.state

        if (!statePolls[state]) {
          statePolls[state] = []
        }

        const poll: ScrapedPoll = {
          pollster: pollData.pollster,
          pollsterSlug: this.normalizePollsterSlug(pollData.pollster),
          pollDate: this.parseDate(pollData.endDate) || new Date(),
          fieldDateStart: this.parseDate(pollData.startDate),
          fieldDateEnd: this.parseDate(pollData.endDate),
          sampleSize: this.parseNumber(pollData.sampleSize),
          methodology: this.mapMethodology(pollData.methodology),
          populationType: this.mapPopulation(pollData.population),
          results: pollData.results,
          sourceUrl: `${this.baseUrl}/polls/`,
        }

        statePolls[state].push(poll)
      }
    }

    return statePolls
  }

  /**
   * Map methodology strings to standard format
   */
  private mapMethodology(method: string): 'phone' | 'online' | 'ivr' | 'sms' | 'mixed' | undefined {
    if (!method) return undefined

    const lower = method.toLowerCase()

    if (lower.includes('phone') || lower.includes('live')) return 'phone'
    if (lower.includes('online') || lower.includes('web')) return 'online'
    if (lower.includes('ivr') || lower.includes('automated')) return 'ivr'
    if (lower.includes('sms') || lower.includes('text')) return 'sms'
    if (lower.includes('mixed') || lower.includes('multi')) return 'mixed'

    return undefined
  }

  /**
   * Map population type to standard format
   */
  private mapPopulation(pop: string): 'rv' | 'lv' | 'a' | undefined {
    if (!pop) return undefined

    const lower = pop.toLowerCase()

    if (lower.includes('lv') || lower.includes('likely')) return 'lv'
    if (lower.includes('rv') || lower.includes('registered')) return 'rv'
    if (lower.includes('a') || lower.includes('adult')) return 'a'

    return undefined
  }

  /**
   * Extract unique candidates from polls
   */
  private extractCandidates(polls: ScrapedPoll[]): string[] {
    const candidateSet = new Set<string>()

    polls.forEach((poll) => {
      Object.keys(poll.results).forEach((candidate) => {
        // Filter out non-candidate answers
        if (
          !candidate.toLowerCase().includes('undecided') &&
          !candidate.toLowerCase().includes('other') &&
          !candidate.toLowerCase().includes('neither')
        ) {
          candidateSet.add(candidate)
        }
      })
    })

    return Array.from(candidateSet)
  }
}
